# Homework #3, Question #2:

In [1]:
# The following line will import LinearRegression Class. 

from sklearn.linear_model import LinearRegression

# Randomly splitting the original dataset into training set and testing set
# The function"train_test_split" from "sklearn.cross_validation" library performs random splitting.
#For this question the Testing Set will be 20% and the Training Set will be 80%

from sklearn.model_selection import train_test_split

# from sklearn.preprocessing import scale

#This line of code will import the preprocessing object
from sklearn import preprocessing


from sklearn import metrics

In [2]:
# Importing the required packages and libraries
# we will need numpy and pandas later

import numpy as np
import pandas as pd

# For part A and B:

In [3]:
# read the dataset as a CSV file directly from the book website
credit_data = pd.read_csv('http://www-bcf.usc.edu/~gareth/ISL/Credit.csv', index_col=0)

# show the first 5 rows of the Credit Dataset
credit_data.head()

Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1   14.891   3606     283      2   34         11    Male      No     Yes   
2  106.025   6645     483      3   82         15  Female     Yes     Yes   
3  104.593   7075     514      4   71         11    Male      No      No   
4  148.924   9504     681      3   36         11  Female      No      No   
5   55.882   4897     357      2   68         16    Male      No     Yes   

   Ethnicity  Balance  
1  Caucasian      333  
2      Asian      903  
3      Asian      580  
4      Asian      964  
5  Caucasian      331

# For Part C:

# Here I generate the Vector X and Target y:

In [4]:
# Creating the Feature Matrix:

# create a python list of feature names that appear in the dataset:
feature_cols = ['Income', 'Limit', 'Rating','Cards','Age','Education','Gender','Student','Married','Ethnicity']

# use the above list to select the features:
X = credit_data[feature_cols]

# check the size:
print(X.shape)

# show the first 5 rows of the dataset:
X.head()



(400, 10)


Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1   14.891   3606     283      2   34         11    Male      No     Yes   
2  106.025   6645     483      3   82         15  Female     Yes     Yes   
3  104.593   7075     514      4   71         11    Male      No      No   
4  148.924   9504     681      3   36         11  Female      No      No   
5   55.882   4897     357      2   68         16    Male      No     Yes   

   Ethnicity  
1  Caucasian  
2      Asian  
3      Asian  
4      Asian  
5  Caucasian

In [5]:
# select the target (last column) from the DataFrame
y = credit_data['Balance']

# checking the size of target vector:
y.size

400

# Using OneHotEncoding (On non-numerical Features):

In [6]:
#Used OneHotEncoding on Non-Numerical Features

list_nonNumerical = ['Gender', 'Student', 'Married','Ethnicity']

for column in list_nonNumerical:
    dummies = pd.get_dummies(credit_data[column])
    credit_data[dummies.columns] = dummies
    
getnewFeatures_X = credit_data

# #get first 5 rows of new table
getnewFeatures_X.head()

# getnewFeatures_X.columns

Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1   14.891   3606     283      2   34         11    Male      No     Yes   
2  106.025   6645     483      3   82         15  Female     Yes     Yes   
3  104.593   7075     514      4   71         11    Male      No      No   
4  148.924   9504     681      3   36         11  Female      No      No   
5   55.882   4897     357      2   68         16    Male      No     Yes   

   Ethnicity  Balance   Male  Female  No  Yes  African American  Asian  \
1  Caucasian      333      1       0   0    1                 0      0   
2      Asian      903      0       1   0    1                 0      1   
3      Asian      580      1       0   1    0                 0      1   
4      Asian      964      0       1   1    0                 0      1   
5  Caucasian      331      1       0   0    1                 0      0   

   Caucasian  
1          1  
2          0  
3          0  
4          0  
5          1

# After using OneHotEncoding, now I generate my new Vector X (It includes the dummy columns) and the Target Y(same target): 

In [7]:
# Creating new Feature Matrix:

# create a python list of feature names that appear in the dataset(including dummy columns):
new_feature_cols = ['Income', 'Limit', 'Rating', 'Cards', 'Age', 'Education', ' Male', 'Female', 'No', 'Yes','African American', 'Asian', 'Caucasian']

# use the above list to select the features:
X_2 = getnewFeatures_X[new_feature_cols]

# check the size:
print(X_2.shape)

# show the first 5 rows of the dataset:
X_2.head()

(400, 13)


Income  Limit  Rating  Cards  Age  Education   Male  Female  No  Yes  \
1   14.891   3606     283      2   34         11      1       0   0    1   
2  106.025   6645     483      3   82         15      0       1   0    1   
3  104.593   7075     514      4   71         11      1       0   1    0   
4  148.924   9504     681      3   36         11      0       1   1    0   
5   55.882   4897     357      2   68         16      1       0   0    1   

   African American  Asian  Caucasian  
1                 0      0          1  
2                 0      1          0  
3                 0      1          0  
4                 0      1          0  
5                 0      0          1

# After applying OneHotEncoding, now I Normalize my Data:

In [8]:
#Normalize my new features: 

normalized_X = preprocessing.scale(X_2)

print(normalized_X)

[[-0.86158299 -0.48999879 -0.46553881 ..., -0.57350122 -0.58504861
   1.00501256]
 [ 1.72743711  0.82826106  0.82870309 ..., -0.57350122  1.70925967
  -0.99501244]
 [ 1.68675551  1.01478681  1.02931059 ..., -0.57350122  1.70925967
  -0.99501244]
 ..., 
 [ 0.35946155 -0.24491264 -0.21963285 ..., -0.57350122 -0.58504861
   1.00501256]
 [-0.21280808 -0.95891584 -1.05441888 ..., -0.57350122 -0.58504861
   1.00501256]
 [-0.75334493  0.34199278  0.38866085 ..., -0.57350122  1.70925967
  -0.99501244]]


# For Part D:

# Split my dataset into Training and Testing:

In [9]:
# Splitting the dataset into testing and training:
X_train, X_test, y_train, y_test = train_test_split(normalized_X, y, test_size=0.20, random_state=2)

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(320, 13)
(320,)
(80, 13)
(80,)


# For Part E:

In [10]:
# In the following line, "my_linreg" is instantiated as an "object" of LinearRegression "class". 

my_linreg = LinearRegression()

# fitting the model to the training data:
my_linreg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

# Now I check the coefficients: 

In [11]:
# printing Theta0 using attribute "intercept_":
print(my_linreg.intercept_)

# printing [Theta1, Theta2, Theta3, etc..] using attribute "coef_":
print(my_linreg.coef_)

523.90108029
[-267.02217986  200.48448409  409.50052155   15.16944724  -15.67145052
    6.10273034   -0.82140619    0.82140619   13.06042561  -13.06042561
  -10.97499135   16.41353146   -4.8353767 ]


# Thus, the most important feature is 'Rating' and the least important feature is the 'Age' column.

# For Part F:

In [13]:
# make predictions on the testing set
y_prediction = my_linreg.predict(X_test)

print(y_prediction)

[  166.54361993   691.95419972  1042.33105298   595.95144506   996.2642987
   659.85353709   894.97534817   259.70415765   961.17013478  1100.07006672
   552.13026292   387.40390817   770.31673241   114.4325947    878.1050115
   192.27016771  1381.98534859  -147.81751799   984.3366008    761.56099637
   583.36552055   119.3388829     69.33619521   -85.90137261   472.73770461
  -156.3649882    335.15312242   621.60861886   901.11984474   888.18209396
  -177.00294666   414.35894245    27.76216268  -296.02008804   970.77118888
  1214.45914494   379.47763519   972.70650059  1740.81159744   832.99958227
   778.56827684   497.00408886   367.18245283   439.80838115   657.47843816
   480.42979859   577.83365738  1090.81265603   506.69685542   664.01099306
   789.41594834   304.60705029  1520.14513552  1168.47448252   342.21498996
   844.3661493   -163.92828617   609.92616115   253.05466144   818.25020407
   373.98589073   558.71865757   -29.36809647   961.77925811  -203.20212513
   730.7832966

In [14]:
# Calculating "Mean Square Error" (MSE):
mse = metrics.mean_squared_error(y_test, y_prediction)

# Using numpy sqrt function to take the square root and calculate "Root Mean Square Error" (RMSE)
rmse_balanceTarget = np.sqrt(mse)

print(rmse_balanceTarget)

161.167018297


# For Part G(Cross-Validation):

In [15]:
# importing the method:
from sklearn.cross_validation import cross_val_score

/anaconda/lib/python3.6/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [16]:
# Applying 10-fold cross validation with "linear regression":

#"my_linreg" is instantiated as an "object" of LinearRegression "class". 
my_linreg = LinearRegression()

mse_list = cross_val_score(my_linreg, normalized_X, y, cv=10, scoring='neg_mean_squared_error')

print(mse_list)

[-24538.79942323 -32419.42698688 -34930.46771829 -38281.6079232
 -14219.8846142  -32897.14222102 -30987.6686748  -12812.87194686
 -20679.85048027 -24167.52894531]


In [17]:
# In order to calculate root mean square error (rmse), I made my mse into positive values: 
mse_list_positive = -mse_list

# using numpy sqrt function to calculate rmse:
rmse_list = np.sqrt(mse_list_positive)
print(rmse_list)

[ 156.64864961  180.05395577  186.89694411  195.6568627   119.24715768
  181.37569358  176.03314652  113.1939572   143.80490423  155.45909091]


In [18]:
# Calculated the average RMSE as final result of cross validation:
print(rmse_list.mean())

160.83703623


# Thus the average RMSE equals 160.837 when using Cross-Validation on Linear Regression